{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "authorship_tag": "ABX9TyOkx1oZZYq4e1S9/G50Mo99",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU",
    "gpuClass": "standard"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/holehouse-lab/ALBATROSS-colab/blob/main/idrome_constructor/idrome_constructor.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<h1>IDRome constructor</h1>\n",
        "\n",
        "* Version 1.0 (initial release, May 27th 2023)\n",
        "\n",
        "This notebook enables a complete IDR-ome annotation to be generated from an input FASTA file.\n",
        "\n",
        "Specifically, by uploading a FASTA file this notebook will:\n",
        "\n",
        "1. Predict all IDRs\n",
        "2. Calculate sequence properties for each IDR\n",
        "3. Predict ensemble properties using ALBATROSS\n",
        "4. Return a CSV file with all this information for easy exploration.\n",
        "\n",
        "\n",
        "### Input:\n",
        "The only input file required is a valid FASTA file where each sequence has a unique FASTA header. If the FASTA file was obatined from UniProt then checking the \"Uniprot_Input\" box will enable the UniProt ID to excised into its own column. Note that FASTA headers have columns replaced by ';' so that a bona fide CSV file can be generated without commas messing up column definitions.\n",
        "\n",
        "### Output:\n",
        "Once the notebook is complete, a CSV file called `IDRome_all.csv` will be downloaded.\n",
        "\n",
        "### Performance:\n",
        "If GPU credits are available, the human proteome takes ~1 minute. If no GPU credits are available the human proteome takes more like 6-7 minutes. "
      ],
      "metadata": {
        "id": "tdl19QtXTPvV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Setup \n",
        "#@markdown Run this cell to setup the notebook. This only needs to be done once, and multiple FASTA files can be analyzed using the cell below once the setup has been run.\n",
        "!pip install git+https://git@github.com/holehouse-lab/shephard.git --quiet;\n",
        "!pip install git+https://git@github.com/idptools/sparrow.git@albatross_v2 --quiet;\n",
        "!pip install git+https://git@github.com/idptools/metapredict.git@batch_mode --quiet;\n",
        "\n",
        "from google.colab import files\n",
        "import io\n",
        "import protfasta\n",
        "\n",
        "from sparrow import Protein\n",
        "from shephard.apis import fasta, uniprot\n",
        "from shephard.apis import metapredict\n",
        "from sparrow.predictors import batch_predict\n",
        "import numpy as np"
      ],
      "metadata": {
        "id": "UPf-D9eYxmYA",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "odOIh46rngCJ",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Run predictions\n",
        "#@markdown <h1>Input options</h1> \n",
        "\n",
        "#@markdown <br>\n",
        "# define the function that will be called when the form is submitted\n",
        "\n",
        "#@markdown <h3>UniProt-generated FASTA file</h3> \n",
        "#@markdown If `uniprot_input` is selected then this analysis assumes \n",
        "#@markdown the passed fasta file was generated by UniProt and will parse\n",
        "#@markdown out the UniProt ID into its own column. If not, the whole FASTA header\n",
        "#@markdown will be used as the unique ID\n",
        "\n",
        "uniprot_input = False #@param {type:\"boolean\"}\n",
        "#@markdown <br>\n",
        "\n",
        "#@markdown <h3>Ensemble properties</h3> \n",
        "#@markdown By default all possible ensemble properties are predicted, although if you\n",
        "#@markdown don't want specific ones to to be predicted this can be adjusted.\n",
        "radius_of_gyration = True #@param {type:\"boolean\"}\n",
        "end_to_end_distance = True #@param {type:\"boolean\"}\n",
        "asphericity = True #@param {type:\"boolean\"}\n",
        "scaling_exponent = True #@param {type:\"boolean\"}\n",
        "prefactor = True #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "\n",
        "# upload FASTA file\n",
        "uploaded_data = files.upload()\n",
        "uploaded_fasta = list(uploaded_data.keys())[0]\n",
        "\n",
        "# read in FASTA file\n",
        "if uniprot_input:\n",
        "  prot = uniprot.uniprot_fasta_to_proteome(uploaded_fasta, invalid_sequence_action='convert-ignore')\n",
        "else:\n",
        "  prot = fasta.fasta_to_proteome(uploaded_fasta, use_header_as_unique_ID=True, invalid_sequence_action='convert-ignore')\n",
        "\n",
        "# predict IDRs...\n",
        "print('Predicting IDRs...',end='')\n",
        "metapredict.annotate_proteome_with_disordered_domains(prot)\n",
        "print('Disorder prediction done')\n",
        "\n",
        "data = {}\n",
        "for d in prot.domains:\n",
        "  name = f\"{d.protein.unique_ID}_{d.start}_{d.end}\"\n",
        "\n",
        "  name = name.replace(',',';')\n",
        "  data[name] = d.sequence\n",
        "\n",
        "\n",
        "\n",
        "## ------------------------------------------  \n",
        "##\n",
        "## RUN PREDICTIONS\n",
        "##\n",
        "\n",
        "if radius_of_gyration:\n",
        "  print('Predicting radii of gyration')\n",
        "  rg = batch_predict.batch_predict(data, network='scaled_rg')  \n",
        "\n",
        "if end_to_end_distance:\n",
        "  print('Predicting end-to-end distance(s)')\n",
        "  re = batch_predict.batch_predict(data, network='scaled_re')  \n",
        "\n",
        "if asphericity:\n",
        "  print('Predicting asphericities')\n",
        "  asph = batch_predict.batch_predict(data, network='asphericity')  \n",
        "\n",
        "if scaling_exponent:\n",
        "  print('Predict scaling exponent')\n",
        "  nu = batch_predict.batch_predict(data, network='scaling_exponent')  \n",
        "\n",
        "if prefactor:\n",
        "  print('Predict scaling prefactor exponent')\n",
        "  pref = batch_predict.batch_predict(data, network='prefactor')  \n",
        "\n",
        "\n",
        "## ------------------------------------------  \n",
        "outname = 'IDRome_all.csv'\n",
        "try:\n",
        "  os.remove(outname)\n",
        "except Exception:\n",
        "  pass\n",
        "\n",
        "fh = open('IDRome_all.csv','w')\n",
        "\n",
        "out_string = ''\n",
        "out_string += \"IDR ID, \"\n",
        "out_string += \"FASTA header, \"\n",
        "\n",
        "if uniprot_input:\n",
        "  out_string += \"UniProtID, \"\n",
        "\n",
        "out_string += \"IDR start, \"\n",
        "out_string += \"IDR end, \"\n",
        "out_string += \"IDR len, \"\n",
        "\n",
        "if radius_of_gyration:\n",
        "  out_string += \"Rg (A), \"\n",
        "\n",
        "if end_to_end_distance:  \n",
        "  out_string += \"Re (A), \"\n",
        "\n",
        "if asphericity:\n",
        "  out_string += \"asphericity, \"\n",
        "\n",
        "if scaling_exponent:\n",
        "  out_string += \"scaling_exponent, \"\n",
        "\n",
        "if prefactor:\n",
        "  out_string += \"prefactor, \"\n",
        "\n",
        "out_string += \"FCR, \"\n",
        "out_string += \"NCPR, \"\n",
        "out_string += \"kappa, \"\n",
        "out_string += \"frac_negative, \"\n",
        "out_string += \"fract_positive, \"\n",
        "out_string += \"fract_aro, \"\n",
        "out_string += \"fract_pro, \"\n",
        "out_string += \"fract_polar, \"\n",
        "out_string += \"fract_ali, \"\n",
        "out_string += \"sequence\\n\"\n",
        "\n",
        "fh.write(out_string)\n",
        "\n",
        "\n",
        "for d in prot.domains:\n",
        "  name = f\"{d.protein.unique_ID}_{d.start}_{d.end}\"\n",
        "  name = name.replace(',',';')\n",
        "  out_string = ''\n",
        "  if name.find(',') > -1:\n",
        "    raise Exception\n",
        "\n",
        "  fasta_header = d.protein.name\n",
        "  fasta_header = fasta_header.replace(',',';')\n",
        "\n",
        "  out_string += f\"{name}, \"\n",
        "  out_string += f\"{fasta_header}, \"\n",
        "\n",
        "  if uniprot_input:\n",
        "    out_string += f\"{d.protein.unique_ID}, \"\n",
        "    \n",
        "  out_string += f\"{d.start}, \"\n",
        "  out_string += f\"{d.end}, \"\n",
        "  out_string += f\"{len(d.sequence)}, \"\n",
        "\n",
        "  if radius_of_gyration:\n",
        "    out_string += f\"{rg[name][1]:.2f}, \"\n",
        "\n",
        "  if end_to_end_distance:\n",
        "    out_string += f\"{re[name][1]:.2f}, \"\n",
        "\n",
        "  if asphericity:\n",
        "    out_string += f\"{asph[name][1]:.3f}, \"\n",
        "\n",
        "  if scaling_exponent:\n",
        "    out_string += f\"{nu[name][1]:.3f}, \"\n",
        "  \n",
        "  if prefactor:\n",
        "    out_string += f\"{pref[name][1]:.3f}, \"\n",
        "\n",
        "  local_protein = Protein(d.sequence)\n",
        "  out_string += f\"{round(local_protein.FCR,3)}, \"\n",
        "  out_string += f\"{round(local_protein.NCPR,3)}, \"\n",
        "  out_string += f\"{round(local_protein.kappa,3)}, \"\n",
        "  out_string += f\"{round(local_protein.fraction_negative,3)}, \"\n",
        "  out_string += f\"{round(local_protein.fraction_positive,3)}, \"\n",
        "  out_string += f\"{round(local_protein.fraction_aromatic,3)}, \"\n",
        "  out_string += f\"{round(local_protein.fraction_proline,3)}, \"\n",
        "  out_string += f\"{round(local_protein.fraction_polar,3)}, \"\n",
        "  out_string += f\"{round(local_protein.fraction_aliphatic,3)}, \"\n",
        "  out_string += f\"{d.sequence}\\n\"\n",
        "  fh.write(out_string)\n",
        "\n",
        "\n",
        "fh.close()\n",
        "\n",
        "files.download('IDRome_all.csv')\n",
        "  \n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Documentation\n",
        "\n",
        "\n",
        "### FASTA input\n",
        "\n",
        "The input file must be a correctly formatted FASTA file. This means each sequence is defined by a header (which starts with a > character) and then on the NEXT line(s) valid amino acid sequence.\n",
        "\n",
        "We also require FASTA headers to be unique.\n",
        "\n",
        "Invalid amino acids will be as best as they can converted to standard amino acids using the following conversion convention::\n",
        "\n",
        "* \"B\"   -> N\n",
        "* \"U\"   -> C\n",
        "* \"X\"   -> G\n",
        "* \"Z\"   -> Q\n",
        "* \" \" -> \\<empty string> (i.e. a whitespace character)\n",
        "* \"\\*\"   -> \\<empty string>\n",
        "* \"\\-\"   -> \\<empty string>\n",
        "\n",
        "\n",
        "### Note\n",
        "Because we want to guarentee that the output file is a true CSV file any commas in your FASTA headers will be removed and replaced with a semi-colon. You have been warned\n",
        "\n",
        "The order of output sequence predictions is guarenteed to match the input order, and individually invalid sequences will be skipped rather\n",
        "\n",
        "### Help\n",
        "I things go wrong, please don't hesitate to [raise an issue on GitHub](https://github.com/holehouse-lab/ALBATROSS-colab)\n",
        "\n"
      ],
      "metadata": {
        "id": "55YY-Vj57lSr"
      }
    }
  ]
}