{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/holehouse-lab/ALBATROSS-colab/blob/main/idrome_constructor/idrome_constructor.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "tdl19QtXTPvV"
   },
   "source": [
    "<h1>IDRome constructor</h1>\n",
    "\n",
    "* Version 3.0 (released, November 5th 2024)\n",
    "* Version 1.0 (initial release, May 27th 2023)\n",
    "* Version 1.1 (fix for metapredict_api update in SHEPHARD, Jan 31st 2024)\n",
    "\n",
    "This notebook enables a complete IDR-ome annotation to be generated from an input FASTA file.\n",
    "\n",
    "Specifically, by uploading a FASTA file this notebook will:\n",
    "\n",
    "1. Predict all IDRs\n",
    "2. Calculate sequence properties for each IDR\n",
    "3. Predict ensemble properties using ALBATROSS\n",
    "4. Return a CSV file with all this information for easy exploration.\n",
    "\n",
    "\n",
    "### Input:\n",
    "The only input file required is a valid FASTA file where each sequence has a unique FASTA header. If the FASTA file was obatined from UniProt then checking the \"Uniprot_Input\" box will enable the UniProt ID to excised into its own column. Note that FASTA headers have columns replaced by ';' so that a bona fide CSV file can be generated without commas messing up column definitions.\n",
    "\n",
    "### Output:\n",
    "Once the notebook is complete, a CSV file called `IDRome_all.csv` will be downloaded.\n",
    "\n",
    "### Performance:\n",
    "If GPU credits are available, the human proteome takes ~1 minute. If no GPU credits are available the human proteome takes more like 6-7 minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "UPf-D9eYxmYA",
    "outputId": "e76b1a48-1907-40ee-ada7-6fcb527326a9"
   },
   "outputs": [],
   "source": [
    "#@title Setup\n",
    "#@markdown Run this cell to setup the notebook. This only needs to be done once, and multiple FASTA files can be analyzed using the cell below once the setup has been run.\n",
    "!pip install git+https://git@github.com/holehouse-lab/shephard.git --quiet;\n",
    "!pip install git+https://git@github.com/idptools/sparrow.git --quiet;\n",
    "!pip install git+https://git@github.com/idptools/metapredict.git --quiet;\n",
    "\n",
    "from google.colab import files\n",
    "import io\n",
    "import protfasta\n",
    "\n",
    "from sparrow import Protein\n",
    "from shephard.apis import fasta, uniprot\n",
    "from shephard.apis import metapredict_api\n",
    "from sparrow.predictors import batch_predict\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 39
    },
    "id": "odOIh46rngCJ",
    "outputId": "3ae435f3-a5f3-4745-ee18-c776557ce092"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "     <input type=\"file\" id=\"files-05c766f1-921e-4248-bbd3-6b8ac7956c0d\" name=\"files[]\" multiple disabled\n",
       "        style=\"border:none\" />\n",
       "     <output id=\"result-05c766f1-921e-4248-bbd3-6b8ac7956c0d\">\n",
       "      Upload widget is only available when the cell has been executed in the\n",
       "      current browser session. Please rerun this cell to enable.\n",
       "      </output>\n",
       "      <script>// Copyright 2017 Google LLC\n",
       "//\n",
       "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
       "// you may not use this file except in compliance with the License.\n",
       "// You may obtain a copy of the License at\n",
       "//\n",
       "//      http://www.apache.org/licenses/LICENSE-2.0\n",
       "//\n",
       "// Unless required by applicable law or agreed to in writing, software\n",
       "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
       "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
       "// See the License for the specific language governing permissions and\n",
       "// limitations under the License.\n",
       "\n",
       "/**\n",
       " * @fileoverview Helpers for google.colab Python module.\n",
       " */\n",
       "(function(scope) {\n",
       "function span(text, styleAttributes = {}) {\n",
       "  const element = document.createElement('span');\n",
       "  element.textContent = text;\n",
       "  for (const key of Object.keys(styleAttributes)) {\n",
       "    element.style[key] = styleAttributes[key];\n",
       "  }\n",
       "  return element;\n",
       "}\n",
       "\n",
       "// Max number of bytes which will be uploaded at a time.\n",
       "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
       "\n",
       "function _uploadFiles(inputId, outputId) {\n",
       "  const steps = uploadFilesStep(inputId, outputId);\n",
       "  const outputElement = document.getElementById(outputId);\n",
       "  // Cache steps on the outputElement to make it available for the next call\n",
       "  // to uploadFilesContinue from Python.\n",
       "  outputElement.steps = steps;\n",
       "\n",
       "  return _uploadFilesContinue(outputId);\n",
       "}\n",
       "\n",
       "// This is roughly an async generator (not supported in the browser yet),\n",
       "// where there are multiple asynchronous steps and the Python side is going\n",
       "// to poll for completion of each step.\n",
       "// This uses a Promise to block the python side on completion of each step,\n",
       "// then passes the result of the previous step as the input to the next step.\n",
       "function _uploadFilesContinue(outputId) {\n",
       "  const outputElement = document.getElementById(outputId);\n",
       "  const steps = outputElement.steps;\n",
       "\n",
       "  const next = steps.next(outputElement.lastPromiseValue);\n",
       "  return Promise.resolve(next.value.promise).then((value) => {\n",
       "    // Cache the last promise value to make it available to the next\n",
       "    // step of the generator.\n",
       "    outputElement.lastPromiseValue = value;\n",
       "    return next.value.response;\n",
       "  });\n",
       "}\n",
       "\n",
       "/**\n",
       " * Generator function which is called between each async step of the upload\n",
       " * process.\n",
       " * @param {string} inputId Element ID of the input file picker element.\n",
       " * @param {string} outputId Element ID of the output display.\n",
       " * @return {!Iterable<!Object>} Iterable of next steps.\n",
       " */\n",
       "function* uploadFilesStep(inputId, outputId) {\n",
       "  const inputElement = document.getElementById(inputId);\n",
       "  inputElement.disabled = false;\n",
       "\n",
       "  const outputElement = document.getElementById(outputId);\n",
       "  outputElement.innerHTML = '';\n",
       "\n",
       "  const pickedPromise = new Promise((resolve) => {\n",
       "    inputElement.addEventListener('change', (e) => {\n",
       "      resolve(e.target.files);\n",
       "    });\n",
       "  });\n",
       "\n",
       "  const cancel = document.createElement('button');\n",
       "  inputElement.parentElement.appendChild(cancel);\n",
       "  cancel.textContent = 'Cancel upload';\n",
       "  const cancelPromise = new Promise((resolve) => {\n",
       "    cancel.onclick = () => {\n",
       "      resolve(null);\n",
       "    };\n",
       "  });\n",
       "\n",
       "  // Wait for the user to pick the files.\n",
       "  const files = yield {\n",
       "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
       "    response: {\n",
       "      action: 'starting',\n",
       "    }\n",
       "  };\n",
       "\n",
       "  cancel.remove();\n",
       "\n",
       "  // Disable the input element since further picks are not allowed.\n",
       "  inputElement.disabled = true;\n",
       "\n",
       "  if (!files) {\n",
       "    return {\n",
       "      response: {\n",
       "        action: 'complete',\n",
       "      }\n",
       "    };\n",
       "  }\n",
       "\n",
       "  for (const file of files) {\n",
       "    const li = document.createElement('li');\n",
       "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
       "    li.append(span(\n",
       "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
       "        `last modified: ${\n",
       "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
       "                                    'n/a'} - `));\n",
       "    const percent = span('0% done');\n",
       "    li.appendChild(percent);\n",
       "\n",
       "    outputElement.appendChild(li);\n",
       "\n",
       "    const fileDataPromise = new Promise((resolve) => {\n",
       "      const reader = new FileReader();\n",
       "      reader.onload = (e) => {\n",
       "        resolve(e.target.result);\n",
       "      };\n",
       "      reader.readAsArrayBuffer(file);\n",
       "    });\n",
       "    // Wait for the data to be ready.\n",
       "    let fileData = yield {\n",
       "      promise: fileDataPromise,\n",
       "      response: {\n",
       "        action: 'continue',\n",
       "      }\n",
       "    };\n",
       "\n",
       "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
       "    let position = 0;\n",
       "    do {\n",
       "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
       "      const chunk = new Uint8Array(fileData, position, length);\n",
       "      position += length;\n",
       "\n",
       "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
       "      yield {\n",
       "        response: {\n",
       "          action: 'append',\n",
       "          file: file.name,\n",
       "          data: base64,\n",
       "        },\n",
       "      };\n",
       "\n",
       "      let percentDone = fileData.byteLength === 0 ?\n",
       "          100 :\n",
       "          Math.round((position / fileData.byteLength) * 100);\n",
       "      percent.textContent = `${percentDone}% done`;\n",
       "\n",
       "    } while (position < fileData.byteLength);\n",
       "  }\n",
       "\n",
       "  // All done.\n",
       "  yield {\n",
       "    response: {\n",
       "      action: 'complete',\n",
       "    }\n",
       "  };\n",
       "}\n",
       "\n",
       "scope.google = scope.google || {};\n",
       "scope.google.colab = scope.google.colab || {};\n",
       "scope.google.colab._files = {\n",
       "  _uploadFiles,\n",
       "  _uploadFilesContinue,\n",
       "};\n",
       "})(self);\n",
       "</script> "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#@title Run predictions\n",
    "#@markdown <h1>Input options</h1>\n",
    "\n",
    "#@markdown <br>\n",
    "# define the function that will be called when the form is submitted\n",
    "\n",
    "#@markdown <h3>UniProt-generated FASTA file</h3>\n",
    "#@markdown If `uniprot_input` is selected then this analysis assumes\n",
    "#@markdown the passed fasta file was generated by UniProt and will parse\n",
    "#@markdown out the UniProt ID into its own column. If not, the whole FASTA header\n",
    "#@markdown will be used as the unique ID\n",
    "\n",
    "uniprot_input = True #@param {type:\"boolean\"}\n",
    "#@markdown <br>\n",
    "\n",
    "#@markdown <h3>Ensemble properties</h3>\n",
    "#@markdown By default all possible ensemble properties are predicted, although if you\n",
    "#@markdown don't want specific ones to to be predicted this can be adjusted.\n",
    "radius_of_gyration = True #@param {type:\"boolean\"}\n",
    "end_to_end_distance = True #@param {type:\"boolean\"}\n",
    "asphericity = True #@param {type:\"boolean\"}\n",
    "scaling_exponent = True #@param {type:\"boolean\"}\n",
    "prefactor = True #@param {type:\"boolean\"}\n",
    "#@markdown ### Select metapredict version\n",
    "metapredict_version = \"v3\" #@param [\"v1\", \"v2\",\"v3\"]\n",
    "\n",
    "\n",
    "\n",
    "# upload FASTA file\n",
    "uploaded_data = files.upload()\n",
    "uploaded_fasta = list(uploaded_data.keys())[0]\n",
    "\n",
    "# read in FASTA file\n",
    "if uniprot_input:\n",
    "  prot = uniprot.uniprot_fasta_to_proteome(uploaded_fasta, invalid_sequence_action='convert-ignore')\n",
    "else:\n",
    "  prot = fasta.fasta_to_proteome(uploaded_fasta, use_header_as_unique_ID=True, invalid_sequence_action='convert-ignore')\n",
    "\n",
    "# predict IDRs...\n",
    "print('Predicting IDRs...',end='')\n",
    "metapredict_api.annotate_proteome_with_disordered_domains(prot, version=metapredict_version, device='cuda')\n",
    "print('Disorder prediction done')\n",
    "\n",
    "data = {}\n",
    "for d in prot.domains:\n",
    "  name = f\"{d.protein.unique_ID}_{d.start}_{d.end}\"\n",
    "\n",
    "  name = name.replace(',',';')\n",
    "  data[name] = d.sequence\n",
    "\n",
    "\n",
    "\n",
    "## ------------------------------------------\n",
    "##\n",
    "## RUN PREDICTIONS\n",
    "##\n",
    "\n",
    "if radius_of_gyration:\n",
    "  print('Predicting radii of gyration')\n",
    "  rg = batch_predict.batch_predict(data, network='scaled_rg')\n",
    "\n",
    "if end_to_end_distance:\n",
    "  print('Predicting end-to-end distance(s)')\n",
    "  re = batch_predict.batch_predict(data, network='scaled_re')\n",
    "\n",
    "if asphericity:\n",
    "  print('Predicting asphericities')\n",
    "  asph = batch_predict.batch_predict(data, network='asphericity')\n",
    "\n",
    "if scaling_exponent:\n",
    "  print('Predict scaling exponent')\n",
    "  nu = batch_predict.batch_predict(data, network='scaling_exponent')\n",
    "\n",
    "if prefactor:\n",
    "  print('Predict scaling prefactor exponent')\n",
    "  pref = batch_predict.batch_predict(data, network='prefactor')\n",
    "\n",
    "\n",
    "## ------------------------------------------\n",
    "outname = 'IDRome_all.csv'\n",
    "try:\n",
    "  os.remove(outname)\n",
    "except Exception:\n",
    "  pass\n",
    "\n",
    "fh = open('IDRome_all.csv','w')\n",
    "\n",
    "out_string = ''\n",
    "out_string += \"IDR ID, \"\n",
    "out_string += \"FASTA header, \"\n",
    "\n",
    "if uniprot_input:\n",
    "  out_string += \"UniProtID, \"\n",
    "\n",
    "out_string += \"IDR start, \"\n",
    "out_string += \"IDR end, \"\n",
    "out_string += \"IDR len, \"\n",
    "\n",
    "if radius_of_gyration:\n",
    "  out_string += \"Rg (A), \"\n",
    "\n",
    "if end_to_end_distance:\n",
    "  out_string += \"Re (A), \"\n",
    "\n",
    "if asphericity:\n",
    "  out_string += \"asphericity, \"\n",
    "\n",
    "if scaling_exponent:\n",
    "  out_string += \"scaling_exponent, \"\n",
    "\n",
    "if prefactor:\n",
    "  out_string += \"prefactor, \"\n",
    "\n",
    "out_string += \"FCR, \"\n",
    "out_string += \"NCPR, \"\n",
    "out_string += \"kappa, \"\n",
    "out_string += \"fract_negative, \"\n",
    "out_string += \"fract_positive, \"\n",
    "out_string += \"fract_aro, \"\n",
    "out_string += \"fract_pro, \"\n",
    "out_string += \"fract_polar, \"\n",
    "out_string += \"fract_ali, \"\n",
    "out_string += \"sequence\\n\"\n",
    "\n",
    "fh.write(out_string)\n",
    "\n",
    "\n",
    "for d in prot.domains:\n",
    "  name = f\"{d.protein.unique_ID}_{d.start}_{d.end}\"\n",
    "  name = name.replace(',',';')\n",
    "  out_string = ''\n",
    "  if name.find(',') > -1:\n",
    "    raise Exception\n",
    "\n",
    "  fasta_header = d.protein.name\n",
    "  fasta_header = fasta_header.replace(',',';')\n",
    "\n",
    "  out_string += f\"{name}, \"\n",
    "  out_string += f\"{fasta_header}, \"\n",
    "\n",
    "  if uniprot_input:\n",
    "    out_string += f\"{d.protein.unique_ID}, \"\n",
    "\n",
    "  out_string += f\"{d.start}, \"\n",
    "  out_string += f\"{d.end}, \"\n",
    "  out_string += f\"{len(d.sequence)}, \"\n",
    "\n",
    "  if radius_of_gyration:\n",
    "    out_string += f\"{rg[name][1]:.2f}, \"\n",
    "\n",
    "  if end_to_end_distance:\n",
    "    out_string += f\"{re[name][1]:.2f}, \"\n",
    "\n",
    "  if asphericity:\n",
    "    out_string += f\"{asph[name][1]:.3f}, \"\n",
    "\n",
    "  if scaling_exponent:\n",
    "    out_string += f\"{nu[name][1]:.3f}, \"\n",
    "\n",
    "  if prefactor:\n",
    "    out_string += f\"{pref[name][1]:.3f}, \"\n",
    "\n",
    "  local_protein = Protein(d.sequence)\n",
    "  out_string += f\"{round(local_protein.FCR,3)}, \"\n",
    "  out_string += f\"{round(local_protein.NCPR,3)}, \"\n",
    "  out_string += f\"{round(local_protein.kappa,3)}, \"\n",
    "  out_string += f\"{round(local_protein.fraction_negative,3)}, \"\n",
    "  out_string += f\"{round(local_protein.fraction_positive,3)}, \"\n",
    "  out_string += f\"{round(local_protein.fraction_aromatic,3)}, \"\n",
    "  out_string += f\"{round(local_protein.fraction_proline,3)}, \"\n",
    "  out_string += f\"{round(local_protein.fraction_polar,3)}, \"\n",
    "  out_string += f\"{round(local_protein.fraction_aliphatic,3)}, \"\n",
    "  out_string += f\"{d.sequence}\\n\"\n",
    "  fh.write(out_string)\n",
    "\n",
    "\n",
    "fh.close()\n",
    "\n",
    "files.download('IDRome_all.csv')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "55YY-Vj57lSr"
   },
   "source": [
    "# Documentation\n",
    "\n",
    "\n",
    "### FASTA input\n",
    "\n",
    "The input file must be a correctly formatted FASTA file. This means each sequence is defined by a header (which starts with a > character) and then on the NEXT line(s) valid amino acid sequence.\n",
    "\n",
    "We also require FASTA headers to be unique.\n",
    "\n",
    "Invalid amino acids will be as best as they can converted to standard amino acids using the following conversion convention::\n",
    "\n",
    "* \"B\"   -> N\n",
    "* \"U\"   -> C\n",
    "* \"X\"   -> G\n",
    "* \"Z\"   -> Q\n",
    "* \" \" -> \\<empty string> (i.e. a whitespace character)\n",
    "* \"\\*\"   -> \\<empty string>\n",
    "* \"\\-\"   -> \\<empty string>\n",
    "\n",
    "\n",
    "### Note\n",
    "Because we want to guarentee that the output file is a true CSV file any commas in your FASTA headers will be removed and replaced with a semi-colon. You have been warned\n",
    "\n",
    "The order of output sequence predictions is guarenteed to match the input order, and individually invalid sequences will be skipped rather\n",
    "\n",
    "### Help\n",
    "If things go wrong, please don't hesitate to [raise an issue on GitHub](https://github.com/holehouse-lab/ALBATROSS-colab)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "include_colab_link": true,
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
