{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMkk5EApLy2LfyRBjGl9vmX",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "gpuClass": "standard",
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/holehouse-lab/EnsPredict-colab/blob/main/example_notebooks/polymer_property_predictors.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Welcome to IDPEnsemble Predictor"
      ],
      "metadata": {
        "id": "ffajaERR2-t_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Install Necessary Dependencies\n",
        "!pip install git+https://git@github.com/idptools/sparrow.git@polymer_predictors --quiet;"
      ],
      "metadata": {
        "id": "-fvNNE364qtj",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "648775a7-456e-4f96-ee8b-9bdd80288fcf",
        "cellView": "form"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m142.6/142.6 kB\u001b[0m \u001b[31m8.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m476.7/476.7 kB\u001b[0m \u001b[31m41.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m101.5/101.5 kB\u001b[0m \u001b[31m13.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.0/58.0 kB\u001b[0m \u001b[31m8.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.3/2.3 MB\u001b[0m \u001b[31m4.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m4.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m86.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Building wheel for sparrow (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for metapredict (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for alphaPredict (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for afrc (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for protfasta (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for getSequence (setup.py) ... \u001b[?25l\u001b[?25hdone\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title File Imports\n",
        "import sparrow\n",
        "from sparrow.predictors.batch_predict import batch_predict\n",
        "from google.colab import data_table\n",
        "import pandas as pd\n",
        "from sparrow.sparrow_exceptions import SparrowException\n",
        "import matplotlib.pyplot as plt\n",
        "from google.colab import files\n",
        "import io\n",
        "from collections import defaultdict\n",
        "import torch"
      ],
      "metadata": {
        "id": "5M6IhZED0ywO",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0125ecb0-0208-4d36-c6fa-6905d9d0833b"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Error importing GPy.\n",
            " If trying to run parrot-optimize, make sure to use `pip install idptools-parrot[optimize]`\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Target Sequence\n",
        "#@markdown <h1>Enter the primary amino acid sequence:</h1>\n",
        "\n",
        "#@markdown Notes: \n",
        "\n",
        "#@markdown - The following predictors implicitly assume this sequence is disordered.\n",
        "\n",
        "#@markdown - A sequence cannot be provided at the same time as a fasta file.\n",
        "\n",
        "# define the function that will be called when the form is submitted\n",
        "def process_form():\n",
        "    uploaded_data = files.upload()\n",
        "    uploaded_fasta = list(uploaded_data.keys())[0]\n",
        "    # Load protein objects\n",
        "    try:\n",
        "      protein_objs = sparrow.read_fasta(uploaded_fasta)\n",
        "    except TypeError as e:\n",
        "      print(f\"Received TypeError: Perhaps you didn't upload a file?\")\n",
        "      raise e\n",
        "\n",
        "    return protein_objs\n",
        "\n",
        "\n",
        "\n",
        "sequence = \"\" #@param {type:\"string\"}\n",
        "fasta_file = True #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "if fasta_file and len(sequence) > 0:\n",
        "  raise SparrowException(f\"Cannot specify both a sequence and a list of sequences from a fasta file simultaneously\")\n",
        "\n",
        "if fasta_file:\n",
        "    protein_objs = process_form()\n",
        "    for protein in protein_objs:\n",
        "        if len(protein) > 750:\n",
        "            raise SparrowException(f\"Sequences longer than 750 residues are not currently supported.\"\\\n",
        "            \"You entered a protein with length {len(protein)} residues\")\n",
        "else:\n",
        "    if len(sequence) > 750:\n",
        "        raise SparrowException(f\"Sequences longer than 750 residues are not currently supported.\"\\\n",
        "            \"You entered a protein with length {len(sequence)} residues\")\n",
        "    \n",
        "    protein_objs = sparrow.Protein(sequence)\n",
        "    "
      ],
      "metadata": {
        "id": "pezBzdsc3RIv",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 76
        },
        "cellView": "form",
        "outputId": "bbe67cb1-747b-40f1-e2e9-efdf5718a88b"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-1dff72d0-6d02-422d-97a6-e1b311260277\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-1dff72d0-6d02-422d-97a6-e1b311260277\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving test.fasta to test.fasta\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Sequence Property Prediction\n",
        "\n",
        "#@markdown <h1>Please select the ensemble features you'd like to predict and then press the run cell button: </h1>\n",
        "\n",
        "#@markdown Predictions can be run on CPU or GPU. Bulk predictions form fasta may be slow on CPU.\n",
        "\n",
        "radius_of_gyration = True #@param {type:\"boolean\"}\n",
        "end_to_end_distance = True #@param {type:\"boolean\"}\n",
        "scaling_exponent = True #@param {type:\"boolean\"}\n",
        "prefactor = True #@param {type:\"boolean\"}\n",
        "asphericity = True #@param {type:\"boolean\"}\n",
        "\n",
        "def __gpu_accelerated_fasta_predictions(protein_objs):\n",
        "    \n",
        "    dfs = []\n",
        "    labels = {header : protein.sequence for header,protein in protein_objs.items()}\n",
        "    lengths = {protein.sequence : int(len(protein)) for header, protein in protein_objs.items()}\n",
        "    \n",
        "    dfs.append(pd.DataFrame(list(labels.items()), columns=['fasta_header', 'sequence']))\n",
        "    dfs.append(pd.DataFrame(list(lengths.items()), columns=['sequence', 'lengths']))\n",
        "    if radius_of_gyration:\n",
        "        radius_of_gyrations = batch_predict(protein_objs,batch_size=32,network=\"rg\")\n",
        "        dfs.append(pd.DataFrame(list(radius_of_gyrations.items()), columns=['sequence', 'radius_of_gyration']))\n",
        "\n",
        "    if end_to_end_distance:\n",
        "        end_to_end_distances = batch_predict(protein_objs,batch_size=32,network=\"re\")\n",
        "        dfs.append(pd.DataFrame(list(end_to_end_distances.items()), columns=['sequence', 'end_to_end_distance']))\n",
        "                \n",
        "    if scaling_exponent:\n",
        "        scaling_exponents = batch_predict(protein_objs,batch_size=32,network=\"scaling_exponent\")\n",
        "        dfs.append(pd.DataFrame(list(scaling_exponents.items()), columns=['sequence', 'scaling_exponent']))\n",
        "\n",
        "    if asphericity:\n",
        "        asphericitys = batch_predict(protein_objs,batch_size=32,network=\"asphericity\") \n",
        "        dfs.append(pd.DataFrame(list(asphericitys.items()), columns=['sequence', 'asphericity']))\n",
        "        \n",
        "    if prefactor:\n",
        "        prefactors = batch_predict(protein_objs,batch_size=32,network=\"prefactor\") \n",
        "        dfs.append(pd.DataFrame(list(prefactors.items()), columns=['sequence', 'prefactor']))\n",
        "\n",
        "    merged_df = dfs[0]  # initialize merged_df with the first dataframe\n",
        "\n",
        "    for df in dfs[1:]:  # iterate over the remaining dataframes\n",
        "        # merge on 'key' column\n",
        "        merged_df = pd.merge(merged_df, df, on='sequence')  \n",
        "        # move 'sequence' column to the end\n",
        "        merged_df = merged_df[[col for col in merged_df.columns if col != 'sequence'] + ['sequence']] \n",
        "    \n",
        "    return merged_df\n",
        "    \n",
        "def __cpu_fasta_predictions(protein_objs):\n",
        "    dimensions = defaultdict(dict)\n",
        "\n",
        "    for i, (header, protein) in enumerate(protein_objs.items(), 1):\n",
        "        dimensions[header].update({\"length\" : int(len(protein))})\n",
        "        dimensions[header].update({\"sequence\" : protein.sequence})\n",
        "\n",
        "        if radius_of_gyration:\n",
        "            dimensions[header].update({\"radius_of_gyration\" : protein.predictor.radius_of_gyration()})\n",
        "        \n",
        "        if end_to_end_distance:\n",
        "            dimensions[header].update({\"end_to_end_distance\" : protein.predictor.end_to_end_distance()})\n",
        "\n",
        "        if scaling_exponent:\n",
        "            dimensions[header].update({\"scaling_exponent\" : protein.predictor.scaling_exponent()})\n",
        "\n",
        "        if asphericity:\n",
        "            dimensions[header].update({\"asphericity\" : protein.predictor.asphericity()})\n",
        "\n",
        "        if prefactor:\n",
        "            dimensions[header].update({\"prefactor\" : protein.predictor.prefactor()})\n",
        "\n",
        "    df = pd.DataFrame(dimensions).T\n",
        "    df = df.rename_axis('fasta_header')\n",
        "    df = df[[col for col in df.columns if col != 'sequence'] + ['sequence']] \n",
        "\n",
        "    return df\n",
        "\n",
        "def return_fasta_predictions(protein_objs):\n",
        "    if torch.cuda.is_available():\n",
        "        df = __gpu_accelerated_fasta_predictions(protein_objs)\n",
        "    else:\n",
        "        df = __cpu_fasta_predictions(protein_objs)\n",
        "        \n",
        "    return df\n",
        "\n",
        "def return_individual_predictions(protein):\n",
        "    dimensions = defaultdict(dict)\n",
        "\n",
        "    dimensions[f\"{protein.sequence}\"].update({\"length\" : int(len(protein))})\n",
        "\n",
        "    if radius_of_gyration:\n",
        "        dimensions[f\"{protein.sequence}\"].update({\"radius_of_gyration\" : protein.predictor.radius_of_gyration()})\n",
        "    \n",
        "    if end_to_end_distance:\n",
        "        dimensions[f\"{protein.sequence}\"].update({\"end_to_end_distance\" : protein.predictor.end_to_end_distance()})\n",
        "    \n",
        "    if scaling_exponent:\n",
        "        dimensions[f\"{protein.sequence}\"].update({\"scaling_exponent\" : protein.predictor.scaling_exponent()})\n",
        "    \n",
        "    if asphericity:\n",
        "        dimensions[f\"{protein.sequence}\"].update({\"asphericity\" : protein.predictor.asphericity()})\n",
        "\n",
        "    if prefactor:\n",
        "        dimensions[f\"{protein.sequence}\"].update({\"prefactor\" : 5.5})\n",
        "\n",
        "    df = pd.DataFrame(dimensions).T\n",
        "    \n",
        "    return df\n",
        "\n",
        "def is_iterable(obj):\n",
        "    try:\n",
        "        iter(obj)\n",
        "        return True\n",
        "    except TypeError:\n",
        "        return False\n",
        "\n",
        "\n",
        "if is_iterable(protein_objs):\n",
        "    df = return_fasta_predictions(protein_objs)\n",
        "else:\n",
        "    df = return_individual_predictions(protein_objs)\n",
        "\n",
        "\n",
        "display(data_table.DataTable(df, include_index=True, num_rows_per_page=5))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 399
        },
        "id": "J5Z377pa3sHV",
        "outputId": "98fb8000-4813-475e-9afc-e76a45fff3ca",
        "collapsed": true
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1/1 [00:00<00:00, 86.55it/s]\n",
            "100%|██████████| 1/1 [00:00<00:00, 59.36it/s]\n",
            "100%|██████████| 1/1 [00:00<00:00, 40.65it/s]\n",
            "100%|██████████| 1/1 [00:00<00:00, 82.30it/s]\n",
            "100%|██████████| 1/1 [00:00<00:00, 80.49it/s]\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>fasta_header</th>\n",
              "      <th>lengths</th>\n",
              "      <th>radius_of_gyration</th>\n",
              "      <th>end_to_end_distance</th>\n",
              "      <th>scaling_exponent</th>\n",
              "      <th>asphericity</th>\n",
              "      <th>prefactor</th>\n",
              "      <th>sequence</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>tr|A0A452YC24|A0A452YC24_AEGTS Inosine triphos...</td>\n",
              "      <td>205</td>\n",
              "      <td>45.207512</td>\n",
              "      <td>106.939102</td>\n",
              "      <td>0.551764</td>\n",
              "      <td>0.418464</td>\n",
              "      <td>6.336755</td>\n",
              "      <td>MSGAAAAVARVLPKAVTFVTGNAKKLEEVRAILGSSIPFQSLKLDL...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>tr|A0A453ALH5|A0A453ALH5_AEGTS Lon protease ho...</td>\n",
              "      <td>994</td>\n",
              "      <td>114.309158</td>\n",
              "      <td>287.563934</td>\n",
              "      <td>0.560979</td>\n",
              "      <td>0.410310</td>\n",
              "      <td>6.261775</td>\n",
              "      <td>PKPERQSTPPAHSRLLVQTLESHRPPSESLMLRAAAAAAAAFPARF...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>tr|A0A453B1F5|A0A453B1F5_AEGTS UMP-CMP kinase ...</td>\n",
              "      <td>214</td>\n",
              "      <td>45.372711</td>\n",
              "      <td>104.782883</td>\n",
              "      <td>0.559540</td>\n",
              "      <td>0.418276</td>\n",
              "      <td>6.591907</td>\n",
              "      <td>WFFVCLMVETLFFMQVVAENMLGDKKVTVVFVLGGPGSGKGTQCSN...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "application/vnd.google.colaboratory.module+javascript": "\n      import \"https://ssl.gstatic.com/colaboratory/data_table/99dac6621f6ae8c4/data_table.js\";\n\n      window.createDataTable({\n        data: [[{\n            'v': 0,\n            'f': \"0\",\n        },\n\"tr|A0A452YC24|A0A452YC24_AEGTS Inosine triphosphate pyrophosphatase OS=Aegilops tauschii subsp. strangulata OX=200361 PE=3 SV=1\",\n{\n            'v': 205,\n            'f': \"205\",\n        },\n{\n            'v': 45.20751190185547,\n            'f': \"45.20751190185547\",\n        },\n{\n            'v': 106.93910217285156,\n            'f': \"106.93910217285156\",\n        },\n{\n            'v': 0.5517642498016357,\n            'f': \"0.5517642498016357\",\n        },\n{\n            'v': 0.4184642434120178,\n            'f': \"0.4184642434120178\",\n        },\n{\n            'v': 6.33675479888916,\n            'f': \"6.33675479888916\",\n        },\n\"MSGAAAAVARVLPKAVTFVTGNAKKLEEVRAILGSSIPFQSLKLDLPELQGEPEDISKEKARMAASQVNGPVLVEDTCLCFNALKGLPGPYIKWFLEKIGHEGLNNLLKAYEDKSAFAMCIFSLALGPEEEPITFVGKTAGKIVPARGPADFGWDPVFQPDGFEQTYAEMPKSEKNQISHRGRALALVKEHFASANYEVQGDGLA\"],\n [{\n            'v': 1,\n            'f': \"1\",\n        },\n\"tr|A0A453ALH5|A0A453ALH5_AEGTS Lon protease homolog, mitochondrial OS=Aegilops tauschii subsp. strangulata OX=200361 PE=3 SV=1\",\n{\n            'v': 994,\n            'f': \"994\",\n        },\n{\n            'v': 114.30915832519531,\n            'f': \"114.30915832519531\",\n        },\n{\n            'v': 287.5639343261719,\n            'f': \"287.5639343261719\",\n        },\n{\n            'v': 0.5609785914421082,\n            'f': \"0.5609785914421082\",\n        },\n{\n            'v': 0.41030970215797424,\n            'f': \"0.41030970215797424\",\n        },\n{\n            'v': 6.261775493621826,\n            'f': \"6.261775493621826\",\n        },\n\"PKPERQSTPPAHSRLLVQTLESHRPPSESLMLRAAAAAAAAFPARFAAVPAVAAAEEMRSPLLRVFGSLRGGRSSVMGRRARFCSSSSGSDSEAAAAAAAEAKAEDAVAAEGEAEGKASSAIVSTNPRPEDFLSVIALPLPHRPLFPGFYMPIYVKDQKLLQALIENRKRSAPYAGAFLVKDEEGTDPSIVASPDSSDNKTIDSLKGKELLKRLHEVGTLAQITSIQGDQVVLLGHRRLRITEMVEEDPLTVKVDHLKENTYNKDDDVIKATSFEVISTLREVLKTSSLWKDHVQTYIQHVGDFNYPRLADFGAAISGANKLLCQEVLEELDVDKRLKLTLELVKKDMEISKLQQAIAKAIEEKISGDQRRYLLNEQLKAIKKELGLETDDKTALSEKFRERIEAKKDKCPPHVLQVIEEELTKLQLLEASSSEFNVTRNYLDWLTVLPWGNYSNENFDVHHAQQILDEDHYGLSDVKERILEFIAVGKLRGTSQGKIICLSGPPGVGKTSIGRSIARALNRQFYRFSVGGLADVAEIKGHRRTYVGAMPGKMVQCLKSVGTANPLVLIDEIDKLGRGHSGDPASALLELLDPEQNINFLDHYLDVPIDLSKVLFVCTANVIDTIPNPLLDRMEIISIAGYITDEKMHIARDYLEKNTREACGIKPEQVEVTDAALLALIENYCREAGVRNLQKQIEKIYRKIALQLVRQGVSNEPALEAVIVTPSVEPTDGGSATIVKDEKLNDPAVKDPEEAKEPSLKETLVTEPALSTDTLSTPEGNKDTDGVKEEAADKAVKKVLIDSSNLDDFVGKPVFHAERIYEQTPVGVVMGLAWTSMGGSTLYIETTKVEEGEGKGALVMTGQLGDVMKESAQIAHTVSRAILRDKEPDNKFFADSKLHLHVPAGATPKDGPSAGCTMITSMLSLAMGKPARKDLAMTGEVTLTGRILPIGGVCFLYSLLILGLLNPSVVYTHAVWSSRTRELAYSPKHQLISLL\"],\n [{\n            'v': 2,\n            'f': \"2\",\n        },\n\"tr|A0A453B1F5|A0A453B1F5_AEGTS UMP-CMP kinase OS=Aegilops tauschii subsp. strangulata OX=200361 PE=3 SV=1\",\n{\n            'v': 214,\n            'f': \"214\",\n        },\n{\n            'v': 45.372711181640625,\n            'f': \"45.372711181640625\",\n        },\n{\n            'v': 104.78288269042969,\n            'f': \"104.78288269042969\",\n        },\n{\n            'v': 0.5595395565032959,\n            'f': \"0.5595395565032959\",\n        },\n{\n            'v': 0.41827628016471863,\n            'f': \"0.41827628016471863\",\n        },\n{\n            'v': 6.591907024383545,\n            'f': \"6.591907024383545\",\n        },\n\"WFFVCLMVETLFFMQVVAENMLGDKKVTVVFVLGGPGSGKGTQCSNIVEHFGFTHLSAGDLLRAEIKSGSENGTMIENMIKEGKIVPSEVTIKLLQQAMIKNENDKFLIDGFPRNEENRAAFENVTKISPAFVLFFNCSEEEMERRLLGRNEGRVDDNIETIRKRFKVFVESSLPVIEYYDAKDKVKKIDAAKPIPEVFEDVKAIFAPYAKVIV\"]],\n        columns: [[\"number\", \"index\"], [\"string\", \"fasta_header\"], [\"number\", \"lengths\"], [\"number\", \"radius_of_gyration\"], [\"number\", \"end_to_end_distance\"], [\"number\", \"scaling_exponent\"], [\"number\", \"asphericity\"], [\"number\", \"prefactor\"], [\"string\", \"sequence\"]],\n        columnOptions: [{\"width\": \"1px\", \"className\": \"index_column\"}],\n        rowsPerPage: 5,\n        helpUrl: \"https://colab.research.google.com/notebooks/data_table.ipynb\",\n        suppressOutputScrolling: true,\n        minimumWidth: undefined,\n      });\n    ",
            "text/plain": [
              "<google.colab.data_table.DataTable object>"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Download the CSV file containing your predictions\n",
        "\n",
        "#@markdown <h1>Name of the CSV file that you'd like to save</h1>\n",
        "\n",
        "filename = \"test\" #@param {type:\"string\"}\n",
        "\n",
        "df.to_csv(f\"{filename}.csv\")\n",
        "\n",
        "files.download(f'{filename}.csv')\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "cellView": "form",
        "id": "FkytLcSGgOsf",
        "outputId": "f61faef2-f814-4e17-ff0b-b6a10fc2e6c2"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_523ebbe2-2ac5-43b8-b5bc-604a757650ce\", \"test.csv\", 2055)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Plot predicted end-to-end-distribution\n",
        "bins = 25 #@param {type:\"integer\"}\n",
        "if is_iterable(protein_objs):\n",
        "    raise SparrowException(\"We don't currently support predicting end-to-end distance distributions from fasta files. Please use the single sequence method in the cell above.\")\n",
        "\n",
        "p_of_Re_R, p_of_Re_P = protein_objs.polymeric.get_predicted_nu_dep_end_to_end_distribution()\n",
        "plt.hist(p_of_Re_R, weights=p_of_Re_P, linewidth=0.5, edgecolor='k',bins=bins)\n",
        "plt.xlabel(r\"End to End Distance ($\\AA$)\")\n",
        "plt.ylabel(\"Probability\")\n",
        "plt.axvline(protein_objs.predictor.end_to_end_distance(), linestyle=\"--\",color='r',label=r\"Mean $R_e$\")\n",
        "plt.legend()\n",
        "plt.show()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 281
        },
        "cellView": "form",
        "id": "e0T5qZGeQ_Er",
        "outputId": "de14b41f-2e19-4c12-ef33-35ed219c9445"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SparrowException",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mSparrowException\u001b[0m                          Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-5-5bfae5baa23f>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mbins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m25\u001b[0m \u001b[0;31m#@param {type:\"integer\"}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mis_iterable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotein_objs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mSparrowException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"We don't currently support predicting end-to-end distance distributions from fasta files. Please use the single sequence method in the cell above.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mp_of_Re_R\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp_of_Re_P\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprotein_objs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpolymeric\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_predicted_nu_dep_end_to_end_distribution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mSparrowException\u001b[0m: We don't currently support predicting end-to-end distance distributions from fasta files. Please use the single sequence method in the cell above."
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "8vn22wStdpJZ"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}