{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/holehouse-lab/ALBATROSS-colab/blob/main/example_notebooks/polymer_property_predictors.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ffajaERR2-t_"
   },
   "source": [
    "# Welcome to ALBATROSS-Colab - The IDR global dimensions predictor\n",
    "### Notebook version 1.0\n",
    "\n",
    "**ALBATROSS**; **A** deep-**L**earning **B**ased **A**pproach for predic**T**ing p**R**operties **O**f di**S**ordered protein**S**\n",
    "\n",
    "An easy to use predictor for conformational ensembles of disordered proteins.\n",
    "\n",
    "For more details please refer to our preprint: [INSERT_PREPRINT_HERE](https://)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "rvyrUlJKyUp5"
   },
   "source": [
    "# Instructions:\n",
    "\n",
    "**Quickstart**\n",
    "\n",
    "*   Enter a sequence into the `sequence` box or upload a FASTA file of protein sequences from your computer.\n",
    "*   Select the properties you'd like to generate predictions for\n",
    "*   Select `Runtime` and press `Run all`\n",
    "*   Profit.\n",
    "\n",
    "\n",
    "Troubleshooting\n",
    "\n",
    "\n",
    "*   To leverage GPUs for bulk FASTA file predictions, Check that the runtime type is set to GPU by clicking on the`Runtime` dropdown menu and change the runtime type by selecting `Change runtime type` to work with GPUs.\n",
    "*   If GPU support is still not working, please try clicking `Runtime` -> `Restart Runtime`.\n",
    "\n",
    "\n",
    "Limitations:\n",
    "\n",
    "*   ALBATROSS was trained on simulations of disordered proteins with the coarse-grained force field Mpipi-GG. While Mpipi-GG shows **strong** correlation with experiments, predictions generated by ALBATROSS are limited by the resolution of the underlying force field - i.e., true ensemble dimensions will vary from ALBATROSS predictions for proteins containing secondary structure.\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "-fvNNE364qtj",
    "outputId": "d7cbb140-5f76-43fb-ff67-d4a1653a48a3"
   },
   "outputs": [],
   "source": [
    "#@title Install Necessary Dependencies\n",
    "!pip install git+https://git@github.com/idptools/parrot --quiet;\n",
    "!pip install git+https://git@github.com/idptools/sparrow.git@polymer_predictors --quiet;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "5M6IhZED0ywO",
    "outputId": "2b05f875-317d-452b-f41c-bdf4caa2605e"
   },
   "outputs": [],
   "source": [
    "#@title File Imports\n",
    "import sparrow\n",
    "from sparrow.predictors.batch_predict import batch_predict\n",
    "from google.colab import data_table\n",
    "import pandas as pd\n",
    "from sparrow.sparrow_exceptions import SparrowException\n",
    "\n",
    "# plotting stuff\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "\n",
    "from google.colab import files\n",
    "import io\n",
    "import os\n",
    "from collections import defaultdict\n",
    "import torch\n",
    "\n",
    "# Some plotting formatting stuff for mpl\n",
    "# set to define axes linewidths\n",
    "matplotlib.rcParams['axes.linewidth'] = 0.5\n",
    "\n",
    "# this defines some prefactors so inline figures look nice\n",
    "# on a retina macbook. These can be commented out without any\n",
    "# issue and are solely asthetic.\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format='retina'\n",
    "\n",
    "# UPDATE 2020-12-31 (my preferred font is Avenir...)\n",
    "font = {'family' : 'arial',\n",
    "    \t'weight' : 'normal'}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "pezBzdsc3RIv"
   },
   "outputs": [],
   "source": [
    "#@title Target Sequence\n",
    "#@markdown <h1>Enter the primary amino acid sequence:</h1>\n",
    "\n",
    "#@markdown Notes: \n",
    "\n",
    "#@markdown - The following predictors implicitly assume this sequence is disordered.\n",
    "\n",
    "#@markdown - A sequence cannot be provided at the same time as a fasta file.\n",
    "\n",
    "# define the function that will be called when the form is submitted\n",
    "def process_form():\n",
    "    uploaded_data = files.upload()\n",
    "    uploaded_fasta = list(uploaded_data.keys())[0]\n",
    "    # Load protein objects\n",
    "    try:\n",
    "      protein_objs = sparrow.read_fasta(uploaded_fasta)\n",
    "    except TypeError as e:\n",
    "      print(f\"Received TypeError: Perhaps you didn't upload a file?\")\n",
    "      raise e\n",
    "\n",
    "    return protein_objs\n",
    "\n",
    "\n",
    "\n",
    "sequence = \"THISISANAMINAACIDSEQWENCE\" #@param {type:\"string\"}\n",
    "fasta_file = False #@param {type:\"boolean\"}\n",
    "\n",
    "\n",
    "if fasta_file and len(sequence) > 0:\n",
    "  raise SparrowException(f\"Cannot specify both a sequence and a list of sequences from a fasta file simultaneously\")\n",
    "\n",
    "if fasta_file:\n",
    "    protein_objs = process_form()\n",
    "    for protein in protein_objs:\n",
    "        if len(protein) > 750:\n",
    "            raise SparrowException(f\"Sequences longer than 750 residues are not currently supported.\"\\\n",
    "            \"You entered a protein with length {len(protein)} residues\")\n",
    "else:\n",
    "    if len(sequence) > 750:\n",
    "        raise SparrowException(f\"Sequences longer than 750 residues are not currently supported.\"\\\n",
    "            \"You entered a protein with length {len(sequence)} residues\")\n",
    "    \n",
    "    protein_objs = sparrow.Protein(sequence)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 133
    },
    "id": "J5Z377pa3sHV",
    "outputId": "ae81d000-1eb8-4195-d58a-0c8c175dfb15"
   },
   "outputs": [],
   "source": [
    "#@title Sequence Property Prediction\n",
    "\n",
    "#@markdown <h1>Please select the ensemble features you'd like to predict and then press the run cell button: </h1>\n",
    "\n",
    "#@markdown Predictions can be run on CPU or GPU. Bulk predictions from fasta may be slow on CPU.\n",
    "\n",
    "radius_of_gyration = True #@param {type:\"boolean\"}\n",
    "end_to_end_distance = True #@param {type:\"boolean\"}\n",
    "asphericity = False #@param {type:\"boolean\"}\n",
    "scaling_exponent = False #@param {type:\"boolean\"}\n",
    "prefactor = False #@param {type:\"boolean\"}\n",
    "\n",
    "\n",
    "def __gpu_accelerated_fasta_predictions(protein_objs):\n",
    "    \n",
    "    dfs = []\n",
    "    labels = {header : protein.sequence for header,protein in protein_objs.items()}\n",
    "    lengths = {protein.sequence : int(len(protein)) for header, protein in protein_objs.items()}\n",
    "    \n",
    "    dfs.append(pd.DataFrame(list(labels.items()), columns=['fasta_header', 'sequence']))\n",
    "    dfs.append(pd.DataFrame(list(lengths.items()), columns=['sequence', 'lengths']))\n",
    "    if radius_of_gyration:\n",
    "        radius_of_gyrations = batch_predict(protein_objs,batch_size=32,network=\"rg\")\n",
    "        dfs.append(pd.DataFrame(list(radius_of_gyrations.items()), columns=['sequence', 'radius_of_gyration']))\n",
    "\n",
    "    if end_to_end_distance:\n",
    "        end_to_end_distances = batch_predict(protein_objs,batch_size=32,network=\"re\")\n",
    "        dfs.append(pd.DataFrame(list(end_to_end_distances.items()), columns=['sequence', 'end_to_end_distance']))\n",
    "                \n",
    "    if scaling_exponent:\n",
    "        scaling_exponents = batch_predict(protein_objs,batch_size=32,network=\"scaling_exponent\")\n",
    "        dfs.append(pd.DataFrame(list(scaling_exponents.items()), columns=['sequence', 'scaling_exponent']))\n",
    "\n",
    "    if asphericity:\n",
    "        asphericitys = batch_predict(protein_objs,batch_size=32,network=\"asphericity\") \n",
    "        dfs.append(pd.DataFrame(list(asphericitys.items()), columns=['sequence', 'asphericity']))\n",
    "        \n",
    "    if prefactor:\n",
    "        prefactors = batch_predict(protein_objs,batch_size=32,network=\"prefactor\") \n",
    "        dfs.append(pd.DataFrame(list(prefactors.items()), columns=['sequence', 'prefactor']))\n",
    "\n",
    "    merged_df = dfs[0]  # initialize merged_df with the first dataframe\n",
    "\n",
    "    for df in dfs[1:]:  # iterate over the remaining dataframes\n",
    "        # merge on 'key' column\n",
    "        merged_df = pd.merge(merged_df, df, on='sequence')  \n",
    "        # move 'sequence' column to the end\n",
    "        merged_df = merged_df[[col for col in merged_df.columns if col != 'sequence'] + ['sequence']] \n",
    "    \n",
    "    return merged_df\n",
    "    \n",
    "def __cpu_fasta_predictions(protein_objs):\n",
    "    dimensions = defaultdict(dict)\n",
    "\n",
    "    for i, (header, protein) in enumerate(protein_objs.items(), 1):\n",
    "        dimensions[header].update({\"length\" : int(len(protein))})\n",
    "        dimensions[header].update({\"sequence\" : protein.sequence})\n",
    "\n",
    "        if radius_of_gyration:\n",
    "            dimensions[header].update({\"radius_of_gyration\" : protein.predictor.radius_of_gyration()})\n",
    "        \n",
    "        if end_to_end_distance:\n",
    "            dimensions[header].update({\"end_to_end_distance\" : protein.predictor.end_to_end_distance()})\n",
    "\n",
    "        if scaling_exponent:\n",
    "            dimensions[header].update({\"scaling_exponent\" : protein.predictor.scaling_exponent()})\n",
    "\n",
    "        if asphericity:\n",
    "            dimensions[header].update({\"asphericity\" : protein.predictor.asphericity()})\n",
    "\n",
    "        if prefactor:\n",
    "            dimensions[header].update({\"prefactor\" : protein.predictor.prefactor()})\n",
    "\n",
    "    df = pd.DataFrame(dimensions).T\n",
    "    df = df.rename_axis('fasta_header')\n",
    "    df = df[[col for col in df.columns if col != 'sequence'] + ['sequence']] \n",
    "\n",
    "    return df\n",
    "\n",
    "def return_fasta_predictions(protein_objs):\n",
    "    if torch.cuda.is_available():\n",
    "        df = __gpu_accelerated_fasta_predictions(protein_objs)\n",
    "    else:\n",
    "        df = __cpu_fasta_predictions(protein_objs)\n",
    "        \n",
    "    return df\n",
    "\n",
    "def return_individual_predictions(protein):\n",
    "    dimensions = defaultdict(dict)\n",
    "    col_names = []\n",
    "    \n",
    "    dimensions[f\"{protein.sequence}\"].update({\"length\" : int(len(protein))})\n",
    "    col_names.append(\"length\")\n",
    "\n",
    "    if radius_of_gyration:\n",
    "        dimensions[f\"{protein.sequence}\"].update({\"radius_of_gyration\" : protein.predictor.radius_of_gyration()})\n",
    "        col_names.append(\"radius_of_gyration\")\n",
    "    if end_to_end_distance:\n",
    "        dimensions[f\"{protein.sequence}\"].update({\"end_to_end_distance\" : protein.predictor.end_to_end_distance()})\n",
    "        col_names.append(\"end_to_end_distance\")\n",
    "    if scaling_exponent:\n",
    "        dimensions[f\"{protein.sequence}\"].update({\"scaling_exponent\" : protein.predictor.scaling_exponent()})\n",
    "        col_names.append(\"scaling_exponent\")\n",
    "    if asphericity:\n",
    "        dimensions[f\"{protein.sequence}\"].update({\"asphericity\" : protein.predictor.asphericity()})\n",
    "        col_names.append(\"asphericity\")\n",
    "    if prefactor:\n",
    "        dimensions[f\"{protein.sequence}\"].update({\"prefactor\" : protein.predictor.prefactor()})\n",
    "        col_names.append(\"prefactor\")\n",
    "        \n",
    "    df = pd.DataFrame(dimensions).T\n",
    "    df = df.rename_axis('sequence')\n",
    "    df = df.reindex(columns=col_names)\n",
    "\n",
    "\n",
    "    return df\n",
    "\n",
    "def is_iterable(obj):\n",
    "    try:\n",
    "        iter(obj)\n",
    "        return True\n",
    "    except TypeError:\n",
    "        return False\n",
    "\n",
    "\n",
    "if is_iterable(protein_objs):\n",
    "    df = return_fasta_predictions(protein_objs)\n",
    "else:\n",
    "    df = return_individual_predictions(protein_objs)\n",
    "\n",
    "\n",
    "display(data_table.DataTable(df, include_index=True, num_rows_per_page=5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "FkytLcSGgOsf"
   },
   "outputs": [],
   "source": [
    "#@title Download the CSV file containing your predictions\n",
    "\n",
    "#@markdown <h1>Name of the CSV file that you'd like to save</h1>\n",
    "\n",
    "filename = \"sequence.csv\" #@param {type:\"string\"}\n",
    "\n",
    "filename = os.path.splitext(filename)[0]\n",
    "\n",
    "if filename:\n",
    "    df.to_csv(f\"{filename}.csv\")\n",
    "\n",
    "    files.download(f'{filename}.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 453
    },
    "id": "e0T5qZGeQ_Er",
    "outputId": "9058d444-21b8-4baa-f4e2-2111a3f02998"
   },
   "outputs": [],
   "source": [
    "#@title Plot predicted end-to-end-distribution\n",
    "\n",
    "#@markdown <h2> This is currently an experimental feature.<h2> \n",
    "#@markdown <h5> Briefly, we leverage ALBATROSS predictors to predict the scaling exponent and polymeric prefactor for a given sequence and use these predictions and the nu-dependent SAW model first established in [1], although we use an implementation derived in [2]. Predictions will work well for sequences with accurately predicted scaling exponents; however, the current performance of the ALBATROSS scaling exponents for biological sequences is more modest. We are actively working on improvng this performance</h5>\n",
    "\n",
    "fig = figure(None, (3,1.5), dpi=350)\n",
    "bins = 25 #@param {type:\"integer\"}\n",
    "if is_iterable(protein_objs):\n",
    "    raise SparrowException(\"We don't currently support predicting end-to-end distance distributions from fasta files. Please use the single sequence method in the cell above.\")\n",
    "\n",
    "p_of_Re_R, p_of_Re_P = protein_objs.polymeric.get_predicted_nu_dep_end_to_end_distribution()\n",
    "plt.hist(p_of_Re_R, weights=p_of_Re_P, linewidth=0.5, edgecolor='k',bins=bins, color='purple', alpha=0.4, label=r'$R_e$ distribution ($\\nu$-SAW)')\n",
    "plt.xlabel(r\"End to End Distance ($\\AA$)\", fontsize=7)\n",
    "plt.ylabel(r\"P($R_e$)\", fontsize=7)\n",
    "\n",
    "plt.xticks(fontsize=7)\n",
    "plt.yticks(fontsize=7)\n",
    "\n",
    "local_re_albatross = protein_objs.predictor.end_to_end_distance()\n",
    "local_re_sawnu = protein_objs.polymeric.get_mean_nudep_saw_end_to_end_distance()\n",
    "\n",
    "plt.axvline(local_re_albatross, linestyle=\"-\",color='k',label=r\"Mean $R_e$ (ALBATROSS)\", linewidth=1)\n",
    "plt.axvline(local_re_sawnu, linestyle=\"--\",color='b',label=r\"Mean $R_e$ ($\\nu$-SAW)\", linewidth=1)\n",
    "if abs(local_re_albatross-local_re_sawnu) > 20:\n",
    "    print('WARNING - looks like the the expected mean end-to-end distances are quite different. This may suggest the scaling exponent is poorly estimated and/or this sequence is poorly described as a flexible homopolymr')\n",
    "\n",
    "plt.legend(frameon=False, fontsize=5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "JkaVwCO8JE9x"
   },
   "source": [
    "# References \n",
    "\n",
    "[1]   Zheng, W., Zerze, G. H., Borgia, A., Mittal, J., Schuler, B., & Best, R. B. (2018). Inferring properties of disordered chains from FRET transfer efficiencies. The Journal of Chemical Physics, 148(12), 123329.\n",
    "\n",
    "[2]   Soranno, A. (2020). Physical basis of the disorder-order transition. Archives of Biochemistry and Biophysics, 685, 108305.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Changelog\n",
    "\n",
    "### 1.0\n",
    "* Version 1.0 released (May 4th 2023)"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "authorship_tag": "ABX9TyMUdM/SlFQEcL4Fwa2aaPBj",
   "include_colab_link": true,
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
